### ISSUE TITLE
Vi mode replaces prompt options

### ISSUE THREAD

**Issue by endrebak** [fish-shell/fish-shell#1988]
Whatever prompt I am using with git info showing, going into `vi mode` with `fish_vi_mode` removes any git info.

I am using fish-HEAD from homebrew so
```
$ fish --version
fish, version unknown
```

Setting the prompt with `fish_config` after having started `fish_vi_mode` does not change the prompt either.

**Comment by endrebak** [comment]
I understand this is an experimental feature and not perfect, just wanted to give a heads up as early as possible.

**Comment by ridiculousfish** [comment]
function fish_vi_mode
        function fish_prompt
            fish_vi_prompt
        end
        set -g fish_key_bindings fish_vi_key_bindings
    end

We should have a good story for how to use a custom prompt with vi mode. Should users define their own fish_vi_prompt?

**Comment by ridiculousfish** [comment]
Perhaps, instead of stomping the prompt in `fish_vi_mode`, the default prompt checks whether vi mode is active and branches to `fish_vi_prompt`. Or maybe we run through the sample prompts and update some/all for vi mode.

**Comment by endrebak** [comment]
The only difference is that there is a mode indicator, so wouldn't it be best to just add that indicator to the end of all the prompts if you are in vi mode?

**Comment by zanchey** [comment]
Yes, that sounds reasonable.

It is a bit harder for people who have customised their prompt as they won't get the changes picked up automatically, but I don't know that there's much we can do about that. Perhaps `functions fish_prompt | grep "\$vi_mode\"`... too evil?

**Comment by ridiculousfish** [comment]
What shall we do about this for 2.2.0, if anything?

**Comment by zanchey** [comment]
I think the ability to include the mode marker at a custom point in the prompt is pretty desirable.

Some options as I see it:
 - Status quo: if you want to use vi mode you use the vi prompt. This is not super good for anyone except those using an unmodified default prompt, but at least you will get a predictable outcome for both new users and those upgrading.
 - Make `fish_vi_mode` modify `fish_right_prompt` instead of `fish_prompt`. This is probably better than the first option because the number of people using a right prompt is probably lower.
 - Modify all the sample prompts to include `(fish_vi_prompt_cm)` in a conditional block of some kind. This will not work if you are an upgrading user, because your prompt makes no reference to the sample prompts once you have selected it.
 - Insert an additional prompt either before or after `fish_prompt` (`fish_bind_mode`?) when in non-default modes (hard to get right and probably not wise to add for 2.2.0 at this stage).
 - Drop the idea of displaying the mode by default (ala zsh), which I think is probably a bad idea.

Of these, I think the option with the lowest impact is putting the mode in `fish_right_prompt` and putting a note in the release notes.

**Comment by leitimmel** [comment]
What if fish displayed an info message on entering vi mode which tells the user:
- that `fish_vi_prompt` must be customized for vi mode instead of `fish_prompt`
- that this message can be removed by unsetting `fish_vi_greeting`?

**Comment by leitimmel** [comment]
@ridiculousfish Would this solve the issue for now?

**Comment by ridiculousfish** [comment]
I'm not opposed to a little message the first time you enter fish_vi_mode. We don't have a very good UI for customizing `fish_vi_prompt` though. I want to think about this a little more.

**Comment by zanchey** [comment]
What about:
 * leaving the default `fish_vi_prompt` in place, similar to the default `fish_prompt` in share/functions
 * adding the indicator to all the sample prompts and teaching the webconfig to save both `fish_prompt` and `fish_vi_prompt`
 * making a note in the release notes about the new function
 * (optionally) adding a note on first entry into `fish_vi_mode`

**Comment by leitimmel** [comment]
Sounds good, however I am probably not the only one who writes his prompt and config by hand. I would keep the message for those.

**Comment by ridiculousfish** [comment]
Another suggestion:

* Support a new function `fish_mode_prompt`, but leave it undefined by default
* At the time we run `fish_prompt`, also run `fish_mode_prompt` and append the output.
* `fish_vi_mode` will define `fish_mode_prompt`, if it's not defined.

This is good because:

* It works automatically with all prompts, including custom prompts.
* Users can redefine `fish_mode_prompt` to output something else, or nothing at all, if they wish to do something custom.

The main disadvantage is that it adds more complexity to the prompt architecture, and we would have to document it.

**Comment by zanchey** [comment]
If you're happy to do it I think that's the best method. It is quite a bit more intrusive though...

**Comment by ridiculousfish** [comment]
I just noticed it looks like vi mode is missing documentation. I don't think we need to document all vi keybindings, but we ought to document how to get into vi mode (and ideally out of it).

**Comment by ridiculousfish** [comment]
I'd appreciate testing / review of my changes in #2121. This eliminates the `fish_vi_prompt` function, and instead implements a new function `fish_mode_prompt` which is prepended (not appended) to the left prompt. A consequence is that custom prompts are no longer overwritten when entering vi mode.

**Comment by ridiculousfish** [comment]
Merged to master as 767742c7e7ee53c8aa7950f928d551c63d69448f and Integration_2.2.0 as 064ad7b9812b85192a5377923084cadee4da03d9

**Comment by OJFord** [comment]
There doesn't seem to be any documentation on this by the way - for anyone googling this, you just need to override `fish_mode_prompt`, e.g. to remove completely:
```fish
function fish_mode_prompt
end function
```

This can be done before or after enabling `fish_vi_mode`.
